src/ImageSharp/Formats/Icon/IconDecoderCore.cs

```diff
@@ -0,0 +1,207 @@
+            if (stream.Read(flag) != PngConstants.HeaderBytes.Length)
+            {
+                break;
+            }
+
+            // Reset the stream position.
+            stream.Seek(-PngConstants.HeaderBytes.Length, SeekOrigin.Current);
+
+            bool isPng = flag.SequenceEqual(PngConstants.HeaderBytes);
+
+            // Decode the frame into a temp image buffer. This is disposed after the frame is copied to the result.
+            Image<TPixel> temp = this.GetDecoder(isPng).Decode<TPixel>(stream, cancellationToken);
+            decodedEntries.Add((temp, isPng, i));
+
+            // Since Windows Vista, the size of an image is determined from the BITMAPINFOHEADER structure or PNG image data
+            // which technically allows storing icons with larger than 256 pixels, but such larger sizes are not recommended by Microsoft.
+            this.Dimensions = new(Math.Max(this.Dimensions.Width, temp.Size.Width), Math.Max(this.Dimensions.Height, temp.Size.Height));
+        }
+
+        ImageMetadata metadata = new();
+        BmpMetadata? bmpMetadata = null;
+        PngMetadata? pngMetadata = null;
+        Image<TPixel> result = new(this.Options.Configuration, metadata, decodedEntries.Select(x =>
+        {
+            ImageFrame<TPixel> target = new(this.Options.Configuration, this.Dimensions);
+            ImageFrame<TPixel> source = x.Image.Frames.RootFrameUnsafe;
+            for (int y = 0; y < source.Height; y++)
+            {
+                source.PixelBuffer.DangerousGetRowSpan(y).CopyTo(target.PixelBuffer.DangerousGetRowSpan(y));
+            }
+
+            // Copy the format specific frame metadata to the image.
+            if (x.IsPng)
+            {
+                if (x.Index == 0)
+                {
+                    pngMetadata = x.Image.Metadata.GetPngMetadata();
+                }
+
+                // Bmp does not contain frame specific metadata.
+                target.Metadata.SetFormatMetadata(PngFormat.Instance, target.Metadata.GetPngFrameMetadata());
+            }
+            else if (x.Index == 0)
+            {
+                bmpMetadata = x.Image.Metadata.GetBmpMetadata();
+            }
+
+            // TODO: The inheriting decoder should be responsible for setting the actual data (FromIconDirEntry)
+            // so we can avoid the protected Field1 and Field2 properties and use strong typing.
+            this.GetFrameMetadata(target.Metadata).FromIconDirEntry(this.Entries[x.Index]);
+
+            x.Image.Dispose();
+
+            return target;
+        }).ToArray());
+
+        // Copy the format specific metadata to the image.
+        if (bmpMetadata != null)
+        {
+            result.Metadata.SetFormatMetadata(BmpFormat.Instance, bmpMetadata);
+        }
+
+        if (pngMetadata != null)
+        {
+            result.Metadata.SetFormatMetadata(PngFormat.Instance, pngMetadata);
+        }
```

Comment:
I think the PNG data may not be the only one. ICO file may have 2 and more PNG data.
So, we should store the PngMetadata in the FrameMetadata instead of the ImageMetadata.

Reply:
They’re just for easy conversation really. See below re bits per pixel. You may want to capture palette data also